docs: add turtlebot4 tutorial

boczekbartek opened the pull request:
## Purpose

Should be merged after: 
- https://github.com/RobotecAI/rai/pull/237
- https://github.com/RobotecAI/rai/pull/239
- https://github.com/RobotecAI/rai/pull/241
- https://github.com/RobotecAI/rai/pull/242

Adds turtlebot tutorial

TODOs:
- [x] Whitlisting ros2 interfaces for the robot is needed in hmi. Refactor the code to allow it. (https://github.com/RobotecAI/rai/issues/269)

## Proposed Changes

Adds a simple tutorial that describes how to setup RAI for the new robot.
Adds recording of the tutorial

## Issues
https://github.com/RobotecAI/rai/issues/236

## Testing

boczekbartek committed:
docs: add turtlebot4 tutorial

simplify launchfile

chore(`turtlebot`): impove launchfile, add rviz and nav2 configs

feat(`turtlebot`): add ros2 interfaces whitelist and improve system
prompt

chore: remove unnecessary TODO

WIP: add ros2 interfaces whitelist to hmi

docs(`turtlebot`): add more examples of interaction

docs(`turtlebot`): add infomation about simulation binary

improve readme

fix humble config

Apply suggestions from code review

Co-authored-by: Adam Dąbrowski <adam.dabrowski@robotec.ai>

readme updates

improve Troubleshooting

rearrange the order

add link to generated whoami

fix typo

add step to ensure whoami pkg is built

update link to binary

chore: add again turtlebot rai example script

remove rviz from turtlebot example

improve readme

vendor agnosticy in Troubleshooting

add minor improvement to readme

chore: remove unnecessary f-string

add link to tutorial video recording

fix: turtlebot demo doesn't use gdnio

chore: reduce verbosity
feat: enhance typing

[system_prompt] Add transform checking before and after nav2 actions to

feat(`turtlebot4`): improve system prompt

rachwalk commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,128 @@
+      ```bash
+      cd rai
+      # for Ubuntu 22.04 jammy and ros2 humlbe
+      unzip -d src/examples/turtlebot4/simulation Turtlebot4_PR1_jammyhumlbe.zip
````
```suggestion
      unzip -d src/examples/turtlebot4/simulation Turtlebot4_PR1_jammyhumble.zip
```

Prolly a typo?

rachwalk commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,138 @@
+
+   ```bash
+   ros2 topic list
+   ```
````
I think the name of the expected ros2 topic should be provided

boczekbartek replied:
https://github.com/RobotecAI/rai/pull/240/commits/a908fa030b95b5e10935f7657624827fa35dad19

rachwalk reviewed (approved):
LGTM

adamdbrw commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,138 @@
+# Turtlebot4 tutorial
+
+The following tutorial will be presented during [ROSCon 2024](https://roscon.ros.org/2024/)
```
This is less interesting information as the tutorial is to have longevity beyond ROSCon, and once guided here, people will already know.

boczekbartek replied:
fixed

adamdbrw commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,138 @@
+
+2. Clone and install [rai](https://github.com/RobotecAI/rai)
+
+   1. Follow setps from [1. Setting up the workspace](https://github.com/RobotecAI/rai?tab=readme-ov-file#1-setting-up-the-workspace)
```
```suggestion
   1. Follow steps from [1. Setting up the workspace](https://github.com/RobotecAI/rai?tab=readme-ov-file#1-setting-up-the-workspace)
```

adamdbrw commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,138 @@
+   1. Follow setps from [1. Setting up the workspace](https://github.com/RobotecAI/rai?tab=readme-ov-file#1-setting-up-the-workspace)
+   2. Then follow steps from [2. Build the project](https://github.com/RobotecAI/rai?tab=readme-ov-file#2-build-the-project)
+
+3. Setup your [OpenAI API](https://openai.com/index/openai-api/). Make sure you have set `OPENAI_API_KEY` variable.
```
Instead, link to RAI vendor-agnostic setup. You can summarize (e.g. follow steps 1-3 of RAI)

boczekbartek replied:
unfortunately this tutorial will not work well with Ollama, because of lack of vision-function calling model..

adamdbrw commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,138 @@
+   4. Create robot's identity. Run the `parse_whoami_package`. This will process the documentation, building
+      it into a vector database, which is used by RAI agent to reason about its identity.
+
+      > **NOTE**: Vector database is created using the OpenAI API. Parsing bigger documents
```
Is this possible with another vendor? 
Provide alternative of vector db download instead of building.

boczekbartek replied:
@adamdbrw Ollama can be used to do embeddings, openai will be used to summarize the image. We could summarize the image with LLava model, but with current config setup user would have to change the model before next step (llava doens't support tool calling) - therefore I would skip it not to overcomplicate

adamdbrw commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,138 @@
+      # you will be asked to press `y` to continue
+      ```
+
+   5. Rebuild the workspace
````
```suggestion
   5. Build the workspace which now includes the new package
```

adamdbrw commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,138 @@
+      colcon build --symlink-install
+      ```
+
+   Congratulations! Your `rai_whoami_node` is configured. In the follwing steps
````
```suggestion
   Congratulations! Your `rai_whoami_node` is configured. In the following steps
```

adamdbrw replied:
Also please double check spelling overall etc.

adamdbrw commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,138 @@
+      ```
+
+   Congratulations! Your `rai_whoami_node` is configured. In the follwing steps
+   your RAI agent will have turtlebot4 personality.
````
```suggestion
   your RAI agent will assume a turtlebot4 "personality".
```

adamdbrw commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,138 @@
+   - `HMI Agent` is responsible for human-robot interaction.
+   - `RAI Node` is responsible for executing tasks on the robot.
+
+   For example you can try such prompt:
```
```suggestion
   For example you can try such prompts:
```

adamdbrw commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,138 @@
+   ```
+
+6. Open you internet browser and go to `localhost:8501`
+7. You can interact with your RAI agent through the chat. On the left you can interact
````
```suggestion
7. Open your internet browser with the address `localhost:8501` to can interact with your RAI agent through the chat. On the left you can communicate
```

adamdbrw commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,138 @@
+       game_launcher:=./src/examples/turtlebot4/simulation/TurtleBot4DemoGamePackage/TurtleBot4Demo.GameLauncher
+   ```
+
+6. Open you internet browser and go to `localhost:8501`
````
```suggestion
```

adamdbrw replied:
Note - explaining purpose of action in the next point

boczekbartek replied:
fixed

adamdbrw commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,138 @@
+   - Spin yourself left by 45 degrees (interaction with the robot using ros2 interfaces) - table with the robotic are shoud be visible in the camera
+   - Use robotic arm to pick up a box from the table (identity and intefaces doesn't allow it)
+   - Drive towards the table (when table is visible, testing ability to interpret camera image and perform actions based on the knowledge)
+   - Drive behind the table (likely will fail in current `rai` version due to limitations in spatial reasoning)
```
```suggestion
```

adamdbrw replied:
Remove the example that is likely to fail

adamdbrw commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,138 @@
+
+### Troubleshooting
+
+_My robot doesn't have the identity._
```
```suggestion
_My robot doesn't have an identity._
```

adamdbrw replied:
Also add how this would look like for the user, e.g seeing an error message etc.

adamdbrw commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,138 @@
+   ros2 service call /rai_whoami_identity_service std_srvs/srv/Trigger
+   ```
+
+2. Make sure you have all required files were generated correctly in the `turtlebot4_whoami` package and have simillar sizes.
````
```suggestion
2. Make sure all required files were generated correctly in the `turtlebot4_whoami` package and have similar sizes to those listed below.
```

adamdbrw commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,138 @@
+   ...
+   ```
+
+   If files above are incorrect please check your `OPENAI_API_KEY` and rerun point 4
````
Vendor independent solution?

boczekbartek replied:
https://github.com/RobotecAI/rai/pull/240/commits/f4d48d5a89c54637af3f103128e65970861cbb7f

rachwalk commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,138 @@
+
+      ```bash
+      cd rai
+      # for Ubuntu 22.04 jammy and ros2 humlbe
````
`cd rai` before rai is cloned

boczekbartek replied:
done

rachwalk commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,138 @@
+5. Run rai agent:
+
+   ```bash
+   ros2 launch ./src/examples/turtlebot4/turtlebot.launch.xml \
+       game_launcher:=./src/examples/turtlebot4/simulation/TurtleBot4DemoGamePackage/TurtleBot4Demo.GameLauncher
````
Ubuntu 22 humble fails with:

```
[rai_whoami_node-1] [INFO] [1729006016.012198943] [rai_whoami_node]: Robot constitution loaded from /home/krachwal/Projects/internal/rai/install/turtlebot4_whoami/share/turtlebot4_whoami/description/robot_constitution.txt
[rai_whoami_node-1] Traceback (most recent call last):
[rai_whoami_node-1]   File "/home/krachwal/Projects/internal/rai/install/rai_whoami/lib/rai_whoami/rai_whoami_node", line 33, in <module>
[rai_whoami_node-1]     sys.exit(load_entry_point('rai-whoami', 'console_scripts', 'rai_whoami_node')())
[rai_whoami_node-1]   File "/home/krachwal/Projects/internal/rai/build/rai_whoami/rai_whoami/rai_whoami_node.py", line 191, in main
[rai_whoami_node-1]     who_am_i_node = WhoAmI()
[rai_whoami_node-1]   File "/home/krachwal/Projects/internal/rai/build/rai_whoami/rai_whoami/rai_whoami_node.py", line 81, in __init__
[rai_whoami_node-1]     self.faiss_index = self._load_documentation()
[rai_whoami_node-1]   File "/home/krachwal/Projects/internal/rai/build/rai_whoami/rai_whoami/rai_whoami_node.py", line 84, in _load_documentation
[rai_whoami_node-1]     faiss_index = FAISS.load_local(
[rai_whoami_node-1]   File "/home/krachwal/.cache/pypoetry/virtualenvs/rai-Qp9irBSo-py3.10/lib/python3.10/site-packages/langchain_community/vectorstores/faiss.py", line 1203, in load_local
[rai_whoami_node-1]     index = faiss.read_index(str(path / f"{index_name}.faiss"))
[rai_whoami_node-1]   File "/home/krachwal/.cache/pypoetry/virtualenvs/rai-Qp9irBSo-py3.10/lib/python3.10/site-packages/faiss/swigfaiss_avx2.py", line 10538, in read_index
[rai_whoami_node-1]     return _swigfaiss_avx2.read_index(*args)
[rai_whoami_node-1] RuntimeError: Error in faiss::FileIOReader::FileIOReader(const char*) at /project/faiss/faiss/impl/io.cpp:67: Error: 'f' failed: could not open /home/krachwal/Projects/internal/rai/install/turtlebot4_whoami/share/turtlebot4_whoami/description/index.faiss for reading: No such file or directory
[ERROR] [rai_whoami_node-1]: process has died [pid 58662, exit code 1, cmd '/home/krachwal/Projects/internal/rai/install/rai_whoami/lib/rai_whoami/rai_whoami_node --ros-args -r __node:=rai_whoami_node --params-file /tmp/launch_params_r1_cikmj'].
```

boczekbartek replied:
I rebased the branch on the newest changes in whoami from development - please try again

boczekbartek commented:
@adamdbrw I rebased this branch on development, refactored the code to fit newest apis from development and answered your comments. I plan to add a video shortly and I think it would be good to merge it soon.  Is it fine with you?

boczekbartek committed:
improve `RaiNode`'s summarizer

Update src/examples/turtlebot4/README.md

Co-authored-by: Paweł Liberadzki <pawel.liberadzki@gmail.com>

feat(`vendors`): allow vendor overwrite for rai_whoami and use local
model in turtlebot example

chore(`whoami`): change poetry scripts to shell scripts

PawelLiberadzki commented on src/examples/turtlebot4/README.md:
````diff
@@ -0,0 +1,207 @@
+
+      ```bash
+      cd rai
+      # for Ubuntu 22.04 jammy and ros2 humlbe
````
```suggestion
      # for Ubuntu 22.04 jammy and ros2 humble
```

PawelLiberadzki commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,207 @@
+
+   You should be able to see a simulation scene.
+
+   > **Tip**: Press 1, 2 or 3 on the keyboard when simulation window to change the
```
These shortcuts does not work for me - is this still available? WSAD camera control works fine.

maciejmajek replied:
It might be caused by a bug in 24.09. Which binary did you use?

boczekbartek committed:
add models configuration flexiblity and other general refactoring

rename whitelist to allowlist

improve hmi system prompt

fix: nav2 humble for turtlebot

improve turtlebot readme w/ rag examples and nav2 install

revert langfuse in config.toml

Co-authored-by: Maciej Majek <46171033+maciejmajek@users.noreply.github.com>

revert vendor overwrite

Update src/examples/turtlebot4/README.md

Co-authored-by: Maciej Majek <46171033+maciejmajek@users.noreply.github.com>

README: Apply suggestions from code review

Co-authored-by: Maciej Majek <46171033+maciejmajek@users.noreply.github.com>

docs: minor styling changes (turtlebot4)

chore: move turtlebot readme to docs/examples

chore: simplified readme

chore: minor adjustments

wip

wip

fix: skipable steps

fix: skip initial streamlit message

improve RAG formating in streamlit

Revert "chore: move turtlebot readme to docs/examples"

This reverts commit 8d500acad2961610f65c31e275201f289201b72d.

docs: mention turtlebot example in `docs/demos`

Revert "improve RAG formating in streamlit"

This reverts commit 5b4ada2ae60d7bcd42129387764713ee35b1a0ad.

change links to better looking binaries

maciejmajek commented on src/rai/rai/cli/rai_cli.py:
```diff
@@ -40,14 +40,21 @@ def parse_whoami_package():
-    llm = get_llm_model(model_type="simple_model")
-    embeddings_model = get_embeddings_model()
+    llm = get_llm_model(model_type="simple_model", vendor=args.vendor)
+    embeddings_model = get_embeddings_model(vendor=args.vendor)
```
This is what config.toml is used for. This should the only source of truth regarding configurating vendors in the framework.

boczekbartek replied:
reverted: https://github.com/RobotecAI/rai/pull/240/commits/3bf785669d351a1c5a47f2d6b4688f4e9adb6232

maciejmajek commented on docs/create_robots_whoami.md:
```diff
@@ -30,11 +30,12 @@ Your robot's `whoami` package serves as a configuration package for the `rai_who
 > [!IMPORTANT]
-> Parsing bigger documents might lead to costs. Embedding model can be configured in
+> Parsing bigger documents with Cloud vendors might lead to costs. Consider using the
+> local `ollama` provider for this task. Embedding model can be configured in
```
So if the embedding model can be configured here, why would someone specify the vendor later?

boczekbartek replied:
https://github.com/RobotecAI/rai/pull/240/commits/3bf785669d351a1c5a47f2d6b4688f4e9adb6232

maciejmajek commented on docs/create_robots_whoami.md:
````diff
@@ -30,11 +30,12 @@ Your robot's `whoami` package serves as a configuration package for the `rai_who
 
 ```shell
-poetry run parse_whoami_package src/examples/panda_whoami
+./scripts/parse_whoami_package.sh src/examples/panda_whoami --vendor ollama
````
```suggestion
./scripts/parse_whoami_package.sh src/examples/panda_whoami
```

boczekbartek replied:
https://github.com/RobotecAI/rai/pull/240/commits/3bf785669d351a1c5a47f2d6b4688f4e9adb6232

maciejmajek commented on config.toml:
```diff
@@ -13,17 +15,17 @@ complex_model = "gpt-4o-2024-08-06"
-use_langfuse = false
-host = "https://cloud.langfuse.com"
+use_langfuse = true
+host = "http://localhost:3000"
```
```suggestion
use_langfuse = false
host = "https://cloud.langfuse.com"
```

maciejmajek commented on pyproject.toml:
```diff
@@ -86,7 +86,3 @@ markers = [
     "billable: marks test as billable  (deselect with '-m \"not billable\"')",
 ]
 addopts = "-m 'not billable' --ignore=src"
-
-[tool.poetry.scripts]
-create_rai_ws = "rai.cli.rai_cli:create_rai_ws"
-parse_whoami_package = "rai.cli.rai_cli:parse_whoami_package"
```
Please revert it or change to sh everywhere

boczekbartek replied:
already done

maciejmajek commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,223 @@
+      - Ubuntu 24.04 & ros2 jazzy: [link](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/ETkT-jvozlpBtuG1piDeqggBRmWl5eylIChc_g0v_EetpA?e=EjcJqe)
+
+      > **NOTE** If you would like to make changes to the simulation and create your
+      > own binary please follow this [README.md](https://github.com/RobotecAI/ROSCon2024Tutorial/README.md)
```
```suggestion
      > [!NOTE]  
      > If you would like to make changes to the simulation and create your own binary please follow this [README.md](https://github.com/RobotecAI/ROSCon2024Tutorial/README.md)
```

maciejmajek commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,223 @@
+
+      > **NOTE**: Be cautious if Cloud vendor was chosen step 3 as `embeddings_model`,
+      > because parsing bigger documents might lead to costs.
+
```
```suggestion
```
I think we can skip that information

maciejmajek commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,223 @@
+   For `embeddings_model` `ollama` vendor works well.
+
+4. Configure `rai_whoami_node` (based on ["Your robot identity in RAI"](https://github.com/RobotecAI/rai/blob/development/docs/create_robots_whoami.md) tutorial):
+
```
```suggestion
> [!TIP]
> Generated files can be downloaded from [here](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/EbPZSEdXYaRGoeecu6oJg6QBsI4ZOe_mrU3uOtOflnIjQg?e=HX8ZHB) unzip them to `src/examples/turtlebot4_whoami/description/generated` with a command:
> `unzip -d src/examples/turtlebot4_whoami/description turtlebot4_whoami_generated.zip`
```

maciejmajek commented on src/examples/turtlebot4/README.md:
```diff
@@ -0,0 +1,223 @@
+      > **NOTE**: generated files can be downloaded from [here](https://robotecai-my.sharepoint.com/:u:/g/personal/bartlomiej_boczek_robotec_ai/EbPZSEdXYaRGoeecu6oJg6QBsI4ZOe_mrU3uOtOflnIjQg?e=HX8ZHB)
+      > unzip them to `src/examples/turtlebot4_whoami/description/generated` with a command:
+      > `unzip -d src/examples/turtlebot4_whoami/description turtlebot4_whoami_generated.zip`
+
```
```suggestion
```
I have added suggestion to mention it as the first step

boczekbartek commented:
I rebased, it is ready to merge

boczekbartek commented:
I squashed the commits - please don's squash them. just merge
